### ISSUE TITLE
Shouldn't data projections be invertible ("injective" / "constructor headed")

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#4437]
Consider:
```agda
data Wrap (A : Set) : Set where
  wrap : A -> Wrap A

unwrap : (A : Set) -> Wrap A -> A
unwrap A (wrap x) = x
```
This fails the injectivity analysis since `varToArg`
https://github.com/agda/agda/blob/fb9ac37fe6bb46f06a5e3d3fc68468edc6e75477/src/full/Agda/TypeChecking/Injectivity.hs#L144-L146
fails __hard__ if the variable (in this case, `x`) isn't a `topLevelArg` (in this case, e.g. `A`) of the function:
https://github.com/agda/agda/blob/fb9ac37fe6bb46f06a5e3d3fc68468edc6e75477/src/full/Agda/TypeChecking/Injectivity.hs#L109-L116
Instead, I suppose, we could fail _softly_ and produce a `UnknownHead`!?
By analogy, we also produce `UnknownHead` if the rhs is irrelevant or a proposition.
I am not sure about the consequences of such a change in the general case, but I also wonder why `UnknownHead` is ok in one case but not the other.

**Comment by andreasabel** [comment]
There is also a bit of strangeness about the data projections created by cubical (!?):
```
Issue4437.wrap-0  :  (A : Set) (d : Wrap A) → A
 is projection like in argument 1 for type Issue4437.Wrap
...
Checking injectivity of Issue4437.wrap-0
```
I speculate these auto-generated projections could be labelled as projection-like already when they are created here (@Saizan ?):
https://github.com/agda/agda/blob/e1cd4d0fee11ee76f4ff4612d5c8e447f37f8689/src/full/Agda/TypeChecking/Rules/Data.hs#L554-L560
But if we go for injectivity for data projections, we could instead define them as injective upon creation.  (An injective function cannot be also projection-like.)
Thoughts on the OP, @UlfNorell, @jespercockx?

**Comment by UlfNorell** [comment]
Projection-likeness can give you exponential performance gains while injectivity only gives you less yellow. I'm not sure it's a good idea to default to the latter. I suppose we could add a pragma to let you choose.

`UnknownHead` should be fine. I added the unknown head logic not too long ago, the hard fail should probably have been changed then.

**Comment by andreasabel** [comment]
> `UnknownHead` should be fine. I added the unknown head logic not too long ago, the hard fail should probably have been changed then.

The relevant simplification of `computeHead` is in PR #4438.  However, as this automatically makes data projections injective and thus no longer projection-like, benchmark proj/Data fails now.


> Projection-likeness can give you exponential performance gains while injectivity only gives you less yellow. I'm not sure it's a good idea to default to the latter. I suppose we could add a pragma to let you choose.

Yes, and possibly set the default to the safer opinion, which would be projection-likeness!?
What would the pragmas be? Something like:
```agda
{-# PROJECTION_LIKE f g h #-}
{-# INVERT f g h #-}
```
Without function names, these could be used as adjacency-pragmas (applying to the next definition).

**Comment by Saizan** [comment]
The projections generated by cubical are not meant to be used directly, so we want projection-likeness

**Comment by andreasabel** [comment]
I suppose then they could be created as projection-like from the start?
At the moment, they are created as regular functions, then subjected to the projection-likeness transformation, which has to change the definition...
Since these projections are created for each field of each constructor of each data type, and not only when `--cubical`is on, this could give a small performance boost.

**Comment by Saizan** [comment]
> I suppose then they could be created as projection-like from the start?

Yes, they could. I think I did it this way because I didn't know how to generate the right projection-like representation, probably just me not being familiar with it though.

**Comment by andreasabel** [comment]
Ok, I give it a try.

**Comment by andreasabel** [comment]
Pushed 6134ea793401465ce428d406e77b04eb74b54d61 that generates data projection directly as projection-like.

**Comment by andreasabel** [comment]
Since 2.6.1 is out, we can restart the discussion on this issue @UlfNorell @Saizan @nad.  I'd say given the arguments so far, it would be worth trying to switch the default from injective to projection-like, i.e., we give projection-likeness priority over injectivity.
If this does not break much, we could go with it, otherwise add a flag and pragma to insist on invertibility.

**Comment by nad** [comment]
> it would be worth trying to switch the default from injective to projection-like

The discussion above led me to believe that this was already the default in most cases. What, exactly, do you propose that we should change?

**Comment by andreasabel** [comment]
The current order is (excluding projections and data projections):
1. Check for injectivity.
2. Check for projection likeness.  Injectivity prevents projection-likeness.

The proposal would be to switch the order, and then projection-like things would not qualify for injectivity.

The current problem is that if we fix the injectivity checker (see https://github.com/agda/agda/issues/4437#issuecomment-584299668), some crucial definitions will be injectivity rather than projection-like, and some crucial benchmarks will fail.
